---
title: "Session 6: Management of Small Populations"
---



# Management of Small Populations

*Session Presenters*

![](images/Presenters6.png)

*Written by Kate Farquharson with input from Catherine Grueber.*

## *Required packages*

```{r, warning=FALSE, message=FALSE}
library(dartRverse)
library(hierfstat)

```

*make sure you have the packages installed, see* [Install dartRverse](install.qmd)

## Load data and explore

The data is a subset of the genome-wide SNPs that are publicly available from the following study, @farquharson_restoring_2022.

**The study aims were to investigate:** 

1) The current genetic diversity of Tasmanian devils across their entire range
2) Whether the genetic diversity of the insurance population is representative of the species

::: callout-note

## Questions
- How many SNPs are there? 
- How many individuals are in the dataset?
- What is the distribution of individuals in populations?

:::

```{r}
load("./data/session6_gl.Rdata")
gl
summary(gl@pop)
```

### Allocate colours to populations

Allocate colours based on site location: East / Central / West Tasmania and captive breeding program.

```{r}
levels(gl@pop)
#west = yellow; central = blue; east = green; insurance population/Maria Island = orange
tascols <- c("skyblue", "#74c476", "skyblue", "#74c476", "#FDD835", "orange", "skyblue",
             "orange", "skyblue", "#FDD835", "skyblue", "#74c476", "#74c476", "#FDD835",
             "#74c476")
```

### Report minor allele frequency (MAF)

::: callout-note

## Questions

- Do you notice any patterns in minor allele frequency?
- Are there differences between the overall MAF distribution and that of individual populations?

:::

```{r}
gl.report.maf(gl)
```

### Visualise population structure by PCoA

Run the PCoA, with points shaded by locations. 

::: callout-note

## Questions
- How would you describe any patterns in population structure? 
- How does the population structure of the insurance population (orange) compare to that of the wild?

:::

```{r}
pcoa <- gl.pcoa(gl)
gl.pcoa.plot(pcoa, gl)

#visualise PCoA using the colour groupings of east/central/west/insurance population
gl.pcoa.plot(pcoa, gl, pt.colors=tascols)
```

## Calculate allelic richness for each population

This calculation requires the hierfstat package, so make sure the  hierfstat is loaded using the library function.

```{r}

#convert the genlight object to genind format
gi <- gl2gi(gl)

#convert genind object to hierfstat format
hfstat <- genind2hierfstat(gi)

#calculate allelic richness
ar <- allelic.richness(hfstat)
names(ar)
ar$min.all # rarefied to 12, the smallest sample size for a genotyped locus
summary(ar$Ar) #gives mean AR for each population
ar <- as.data.frame(ar$Ar)
mean.ar <- colMeans(ar)
```

Calculate a measure of variance for allelic richness, e.g. standard error. Standard error = standard deviation / sqrt(n), where n is the number of genotyped loci

```{r}
sd(ar$Bronte, na.rm=TRUE)/sqrt(nrow(ar) - length(which(is.na(ar$Bronte))))
```

Boxplot of allelic richness per population: 


::: callout-note

## Questions

- Which populations have the highest and the lowest allelic richness? 
- Can you think of any explanation?

:::

```{r}
#first, extend the margins of the graphing window to fit long axis labels
par(mar=c(8,3,3,3))
boxplot(ar, ylab="Allelic richness", las=2, col=tascols)

#order boxplot to group locations by west, central, east, insurance population
ar.ordered <- ar[,c(10, 14, 5, 1, 3, 11, 9, 7, 13, 4, 12, 15, 2, 6, 8)]
tascols.ordered <- tascols[c(10,14,5,1,3,11,9,7,13,4,12,15,2,6,8)]
boxplot(ar.ordered, ylab="Allelic richness", las=2, col=tascols.ordered)
```

## Calculate heterozygosity statistics for each population

The function gl.report.heterozygosity reports the observed, expected, and unbiased heterozygosities and Fis (inbreeding coefficient) by population or the observed heterozygosity for each individual in a genlight object.

Calculate heterozygosity and inbreeding statistics for each population. Note that the plot is ordered by default from lowest to highest heterozygosity.


::: callout-note

## Questions
- Which populations have the highest and lowest heterozygosity?
- Within a population, do you notice any patterns of observed compared to expected heterozygosity from the graph? 
- What about Fis?

:::

```{r}
#use the ? to search the help for a function and see what options can be specified
#?gl.report.heterozygosity
hets <- gl.report.heterozygosity(gl, method="pop",plot.colors.pop = tascols)
```

We can use the same function, but switch the method from "pop" to "ind" to calculate individual-level observed heterozygosity. This can be useful if you are investigating invidiual-level diversity, and may also be useful in assessing the impact of sample quality and missing data on your results.

::: callout-note

## Question
- What are some possible explanations for outliers with low individual observed heterozygosity?

:::

```{r}
ind.hets <- gl.report.heterozygosity(gl, method="ind")
```

## Genetic differentiation by Fst

Calculate pairwise Fsts between populations. 

::: callout-note

## Questions
- Which populations have the largest pairwise Fst value? Where are these populations located? 
- Compare the largest Fst value with the PCoA you made earlier. What do you notice?

:::

```{r}
fsts <- gl.fst.pop(gl, nboots=1000, percent=95)
knitr::kable((fsts$Fsts))
#summary(fsts$Fsts)
```

## *Further Study*

@farquharson_restoring_2022
