<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Session 10: Sex-linked markers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Session10_SexLinkedMarkers_files/libs/clipboard/clipboard.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/quarto.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/popper.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/anchor.min.js"></script>
<link href="Session10_SexLinkedMarkers_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Session10_SexLinkedMarkers_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages">Required packages</a></li>
  <li><a href="#dataset-1---zwzz---the-yellow-tufted-honeyeater" id="toc-dataset-1---zwzz---the-yellow-tufted-honeyeater" class="nav-link" data-scroll-target="#dataset-1---zwzz---the-yellow-tufted-honeyeater">Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#run-filter.sex.linked" id="toc-run-filter.sex.linked" class="nav-link" data-scroll-target="#run-filter.sex.linked">Run <code>filter.sex.linked</code></a></li>
  <li><a href="#run-infer.sex" id="toc-run-infer.sex" class="nav-link" data-scroll-target="#run-infer.sex">Run <code>infer.sex</code></a></li>
  </ul></li>
  <li><a href="#dataset-2---xxxy---the-leadbeaters-possum" id="toc-dataset-2---xxxy---the-leadbeaters-possum" class="nav-link" data-scroll-target="#dataset-2---xxxy---the-leadbeaters-possum">Dataset 2 - XX/XY - The Leadbeater’s possum</a>
  <ul class="collapse">
  <li><a href="#load-data-1" id="toc-load-data-1" class="nav-link" data-scroll-target="#load-data-1">Load data</a></li>
  <li><a href="#run-filter.sex.linked-1" id="toc-run-filter.sex.linked-1" class="nav-link" data-scroll-target="#run-filter.sex.linked-1">Run <code>filter.sex.linked</code></a></li>
  <li><a href="#run-infer.sex-1" id="toc-run-infer.sex-1" class="nav-link" data-scroll-target="#run-infer.sex-1">Run <code>infer.sex</code></a></li>
  </ul></li>
  <li><a href="#exercise-4" id="toc-exercise-4" class="nav-link" data-scroll-target="#exercise-4">EXERCISE</a>
  <ul class="collapse">
  <li><a href="#exercise-data-1---your-own-data" id="toc-exercise-data-1---your-own-data" class="nav-link" data-scroll-target="#exercise-data-1---your-own-data">Exercise data 1 - Your own data</a>
  <ul class="collapse">
  <li><a href="#number-of-sex-linked-markers" id="toc-number-of-sex-linked-markers" class="nav-link" data-scroll-target="#number-of-sex-linked-markers">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid" id="toc-individuals-with-wrong-sexid" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid">2. Individuals with wrong sexID?</a></li>
  <li><a href="#changes-in-pca-before-and-after-removing-the-slm" id="toc-changes-in-pca-before-and-after-removing-the-slm" class="nav-link" data-scroll-target="#changes-in-pca-before-and-after-removing-the-slm">3. Changes in PCA before and after removing the SLM?</a></li>
  <li><a href="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" id="toc-differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" class="nav-link" data-scroll-target="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</a></li>
  </ul></li>
  <li><a href="#exercise-data-2---the-eastern-yellow-robin" id="toc-exercise-data-2---the-eastern-yellow-robin" class="nav-link" data-scroll-target="#exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</a>
  <ul class="collapse">
  <li><a href="#load-data-2" id="toc-load-data-2" class="nav-link" data-scroll-target="#load-data-2">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-1" id="toc-number-of-sex-linked-markers-1" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-1">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-1" id="toc-individuals-with-wrong-sexid-1" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-1">2. Individuals with wrong sexID?</a></li>
  <li><a href="#changes-in-pca-before-and-after-removing-the-slm-1" id="toc-changes-in-pca-before-and-after-removing-the-slm-1" class="nav-link" data-scroll-target="#changes-in-pca-before-and-after-removing-the-slm-1">3. Changes in PCA before and after removing the SLM?</a></li>
  <li><a href="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" id="toc-differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" class="nav-link" data-scroll-target="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</a></li>
  </ul></li>
  <li><a href="#exercise-data-3---bull-shark" id="toc-exercise-data-3---bull-shark" class="nav-link" data-scroll-target="#exercise-data-3---bull-shark">Exercise data 3 - Bull shark</a>
  <ul class="collapse">
  <li><a href="#load-data-3" id="toc-load-data-3" class="nav-link" data-scroll-target="#load-data-3">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-2" id="toc-number-of-sex-linked-markers-2" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-2">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-2" id="toc-individuals-with-wrong-sexid-2" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-2">2. Individuals with wrong sexID?</a></li>
  </ul></li>
  <li><a href="#exercise-data-4---blue-shark" id="toc-exercise-data-4---blue-shark" class="nav-link" data-scroll-target="#exercise-data-4---blue-shark">Exercise data 4 - Blue shark</a>
  <ul class="collapse">
  <li><a href="#load-data-4" id="toc-load-data-4" class="nav-link" data-scroll-target="#load-data-4">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-3" id="toc-number-of-sex-linked-markers-3" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-3">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-3" id="toc-individuals-with-wrong-sexid-3" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-3">2. Individuals with wrong sexID?</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Session 10: Sex-linked markers</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Diana Robledo-Ruiz &amp; Floriaan Devloo-Delva </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="required-packages" class="level1">
<h1>Required packages</h1>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-1_1c98dd5edc3b6ca21602712ef92c5e0c">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.base)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.sexlinked)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dataset-1---zwzz---the-yellow-tufted-honeyeater" class="level1">
<h1>Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Yellow_Tufted_Honeyeater.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Yellow Tufted Honeyeater</figcaption>
</figure>
</div>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-2_1925c7e07301fa118ae30d0784a5d9f9">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"YTH"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>YTH  <span class="co"># Explore the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  1,000 SNPs , size: 50 Mb

    missing data: 166024 (=25.9 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  1000 locus labels
   @loc.all:  1000 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>YTH<span class="sc">@</span>n.loc  <span class="co"># Number of SNPs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(YTH<span class="sc">@</span>ind.names)  <span class="co"># Number of individuals</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 641</code></pre>
</div>
</div>
</section>
<section id="run-filter.sex.linked" class="level2">
<h2 class="anchored" data-anchor-id="run-filter.sex.linked">Run <code>filter.sex.linked</code></h2>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (i.e., genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/ZW.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">ZW/ZZ sex chromosomes</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Set <code>ncores</code> to more than 1 (default) if you have more than 50,000 SNPs, since it could actually slow down the analysis with smaller datasets.</p>
</div>
</div>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-3_e2888e81d93697c19084ca9c54887ab1">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(YTH<span class="sc">@</span>other<span class="sc">$</span>ind.metrics))  <span class="co"># Check that ind.metrics has the necessary columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 12%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 17%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">pop</th>
<th style="text-align: left;">sex</th>
<th style="text-align: left;">sex_original</th>
<th style="text-align: left;">service</th>
<th style="text-align: left;">plate_location</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">Melanops</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A1</td>
</tr>
<tr class="even">
<td style="text-align: left;">W49</td>
<td style="text-align: left;">W49</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">W90</td>
<td style="text-align: left;">W90</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A12</td>
</tr>
<tr class="even">
<td style="text-align: left;">C25</td>
<td style="text-align: left;">C25</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C8</td>
<td style="text-align: left;">C8</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A3</td>
</tr>
<tr class="even">
<td style="text-align: left;">W70</td>
<td style="text-align: left;">W70</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A4</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> YTH, <span class="at">system =</span> <span class="st">"zw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 289 females and 347 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 46 sex-linked loci:
   2 W-linked loci
   33 sex-biased loci
   10 Z-linked loci
   1 ZW gametologs.
And 954 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>How many males and females does the dataset contain?</p>
<p>How many sex-linked loci were found?</p>
</div>
</div>
<p>Now check the output:</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-4_b24b8c3392824a273cb024ea88ec52a6">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>w.linked  <span class="co"># Notice that it says 'w-linked'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  2 SNPs , size: 48.9 Mb

    missing data: 695 (=54.21 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  2 locus labels
   @loc.all:  2 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>z.linked  <span class="co"># Notice that it says 'z-linked'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  10 SNPs , size: 48.9 Mb

    missing data: 268 (=4.18 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  10 locus labels
   @loc.all:  10 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>gametolog</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  1 SNPs , size: 48.9 Mb

    missing data: 13 (=2.03 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  1 locus labels
   @loc.all:  1 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>sex.biased</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  33 SNPs , size: 48.9 Mb

    missing data: 5176 (=24.47 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  33 locus labels
   @loc.all:  33 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>autosomal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 641 genotypes,  954 SNPs , size: 50 Mb

    missing data: 159872 (=26.14 %) scored as NA

 ** Genetic data
   @gen: list of 641 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  641 individual labels
   @loc.names:  954 locus labels
   @loc.all:  954 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 1-540)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(res<span class="sc">$</span>results.table))  <span class="co"># The output table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">index</th>
<th style="text-align: right;">count.F.miss</th>
<th style="text-align: right;">count.M.miss</th>
<th style="text-align: right;">count.F.scored</th>
<th style="text-align: right;">count.M.scored</th>
<th style="text-align: right;">ratio</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">p.adjusted</th>
<th style="text-align: right;">scoringRate.F</th>
<th style="text-align: right;">scoringRate.M</th>
<th style="text-align: left;">w.linked</th>
<th style="text-align: left;">sex.biased</th>
<th style="text-align: right;">count.F.het</th>
<th style="text-align: right;">count.M.het</th>
<th style="text-align: right;">count.F.hom</th>
<th style="text-align: right;">count.M.hom</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">stat.p.value</th>
<th style="text-align: right;">stat.p.adjusted</th>
<th style="text-align: right;">heterozygosity.F</th>
<th style="text-align: right;">heterozygosity.M</th>
<th style="text-align: left;">z.linked</th>
<th style="text-align: left;">zw.gametolog</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">27338051-23-G/A</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">216</td>
<td style="text-align: right;">270</td>
<td style="text-align: right;">73</td>
<td style="text-align: right;">77</td>
<td style="text-align: right;">0.8440653</td>
<td style="text-align: right;">0.3987958</td>
<td style="text-align: right;">0.9728507</td>
<td style="text-align: right;">0.2525952</td>
<td style="text-align: right;">0.2219020</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">73</td>
<td style="text-align: right;">77</td>
<td style="text-align: right;">1.0544475</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">27350703-58-A/T</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">285</td>
<td style="text-align: right;">343</td>
<td style="text-align: right;">1.2031527</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.9861592</td>
<td style="text-align: right;">0.9884726</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">85</td>
<td style="text-align: right;">285</td>
<td style="text-align: right;">258</td>
<td style="text-align: right;">0.0106954</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2478134</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">42055937-67-T/G</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">54</td>
<td style="text-align: right;">59</td>
<td style="text-align: right;">235</td>
<td style="text-align: right;">288</td>
<td style="text-align: right;">1.1214686</td>
<td style="text-align: right;">0.6034024</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.8131488</td>
<td style="text-align: right;">0.8299712</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">22</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">213</td>
<td style="text-align: right;">252</td>
<td style="text-align: right;">0.7234483</td>
<td style="text-align: right;">0.2666319</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.0936170</td>
<td style="text-align: right;">0.1250000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">27340971-37-C/G</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">35</td>
<td style="text-align: right;">73</td>
<td style="text-align: right;">254</td>
<td style="text-align: right;">274</td>
<td style="text-align: right;">0.5177315</td>
<td style="text-align: right;">0.0029461</td>
<td style="text-align: right;">0.0701449</td>
<td style="text-align: right;">0.8788927</td>
<td style="text-align: right;">0.7896254</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">254</td>
<td style="text-align: right;">274</td>
<td style="text-align: right;">1.0785629</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">42071187-64-G/T</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">167</td>
<td style="text-align: right;">157</td>
<td style="text-align: right;">180</td>
<td style="text-align: right;">0.9063543</td>
<td style="text-align: right;">0.5766590</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.5432526</td>
<td style="text-align: right;">0.5187320</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">138</td>
<td style="text-align: right;">164</td>
<td style="text-align: right;">1.4097673</td>
<td style="text-align: right;">0.3735398</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.1210191</td>
<td style="text-align: right;">0.0888889</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">27352295-35-A/T</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">191</td>
<td style="text-align: right;">244</td>
<td style="text-align: right;">98</td>
<td style="text-align: right;">103</td>
<td style="text-align: right;">0.8229837</td>
<td style="text-align: right;">0.2662610</td>
<td style="text-align: right;">0.8551634</td>
<td style="text-align: right;">0.3391003</td>
<td style="text-align: right;">0.2968300</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">89</td>
<td style="text-align: right;">94</td>
<td style="text-align: right;">1.0559200</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.0918367</td>
<td style="text-align: right;">0.0873786</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output consists of a genlight object for each type of loci, plus a results table.</p>
</section>
<section id="run-infer.sex" class="level2">
<h2 class="anchored" data-anchor-id="run-infer.sex">Run <code>infer.sex</code></h2>
<p>This function uses the complete output of function filter.sex.linked (list of 6 objects) to infer the sex of all individuals in the dataset. Specifically, the function uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an <code>agreed sex</code>.</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-5_ff1b0a79814469159f599808a6025054">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">W49</td>
<td style="text-align: left;">W49</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">W90</td>
<td style="text-align: left;">W90</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">C25</td>
<td style="text-align: left;">C25</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C8</td>
<td style="text-align: left;">C8</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">7</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">W70</td>
<td style="text-align: left;">W70</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
IMPORTANT
</div>
</div>
<div class="callout-body-container callout-body">
<p>We created this function with the explicit intent that a human checks the evidence for the <code>agreed sex</code> that do NOT agree for all types of sex-linked loci (denoted as ‘*M’ or ‘*F’). This human can then use their criterion to validate these assignments.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>Can you find individuals for which the <code>agreed sex</code> is uncertain (i.e., has an asterisk “*”)?</p>
</div>
</div>
</section>
</section>
<section id="dataset-2---xxxy---the-leadbeaters-possum" class="level1">
<h1>Dataset 2 - XX/XY - The Leadbeater’s possum</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Leadbeaters_possum.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Leadbeater’s possum</figcaption>
</figure>
</div>
<section id="load-data-1" class="level2">
<h2 class="anchored" data-anchor-id="load-data-1">Load data</h2>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-6_f8e99302ffd29a6cd58941a373f98e0a">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"LBP"</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>LBP  <span class="co"># Explore the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  1,000 SNPs , size: 5.2 Mb

    missing data: 18414 (=4.9 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  1000 locus labels
   @loc.all:  1000 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>LBP<span class="sc">@</span>n.loc  <span class="co"># Number of SNPs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(LBP<span class="sc">@</span>ind.names)  <span class="co"># Number of individuals</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 376</code></pre>
</div>
</div>
</section>
<section id="run-filter.sex.linked-1" class="level2">
<h2 class="anchored" data-anchor-id="run-filter.sex.linked-1">Run <code>filter.sex.linked</code></h2>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/XY.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">XX/XY sex chromosomes</figcaption>
</figure>
</div>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-7_c0b961bdb9e22c0b8094d7d89c3d2716">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(LBP<span class="sc">@</span>other<span class="sc">$</span>ind.metrics))  <span class="co"># Check that ind.metrics has the necessary columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 15%">
<col style="width: 23%">
<col style="width: 20%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">sex</th>
<th style="text-align: left;">pop</th>
<th style="text-align: right;">Year.collected</th>
<th style="text-align: left;">service</th>
<th style="text-align: left;">plate_location</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">2001</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A11</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1999</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A3</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> LBP, <span class="at">system =</span> <span class="st">"xy"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 162 females and 211 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 13 sex-linked loci:
   0 Y-linked loci
   3 sex-biased loci
   9 X-linked loci
   1 XY gametologs.
And 987 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-7-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>How many males and females does the dataset contain?</p>
<p>How many sex-linked loci were found?</p>
</div>
</div>
<p>Now check the output:</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-8_57a9b44f770b7d5a59c3bc651f7dd6b8">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>y.linked  <span class="co"># Notice that it says 'y-linked'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  0 SNPs , size: 4.8 Mb

    missing data: 18414 (=Inf %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  0 locus labels
   @loc.all:  0 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>x.linked  <span class="co"># Notice that it says 'x-linked'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  9 SNPs , size: 4.7 Mb

    missing data: 99 (=2.93 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  9 locus labels
   @loc.all:  9 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>gametolog</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  1 SNPs , size: 4.7 Mb

    missing data: 0 (=0 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  1 locus labels
   @loc.all:  1 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>sex.biased</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  3 SNPs , size: 4.7 Mb

    missing data: 261 (=23.14 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  3 locus labels
   @loc.all:  3 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>autosomal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  987 SNPs , size: 5.3 Mb

    missing data: 18054 (=4.86 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  987 locus labels
   @loc.all:  987 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(res<span class="sc">$</span>results.table))  <span class="co"># The output table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">index</th>
<th style="text-align: right;">count.F.miss</th>
<th style="text-align: right;">count.M.miss</th>
<th style="text-align: right;">count.F.scored</th>
<th style="text-align: right;">count.M.scored</th>
<th style="text-align: right;">ratio</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">p.adjusted</th>
<th style="text-align: right;">scoringRate.F</th>
<th style="text-align: right;">scoringRate.M</th>
<th style="text-align: left;">y.linked</th>
<th style="text-align: left;">sex.biased</th>
<th style="text-align: right;">count.F.het</th>
<th style="text-align: right;">count.M.het</th>
<th style="text-align: right;">count.F.hom</th>
<th style="text-align: right;">count.M.hom</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">stat.p.value</th>
<th style="text-align: right;">stat.p.adjusted</th>
<th style="text-align: right;">heterozygosity.F</th>
<th style="text-align: right;">heterozygosity.M</th>
<th style="text-align: left;">x.linked</th>
<th style="text-align: left;">xy.gametolog</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">28686439-63-T/C</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">162</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">1.301530</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">46</td>
<td style="text-align: right;">69</td>
<td style="text-align: right;">116</td>
<td style="text-align: right;">142</td>
<td style="text-align: right;">0.8165373</td>
<td style="text-align: right;">0.4286518</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.2839506</td>
<td style="text-align: right;">0.3270142</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28685838-47-G/C</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">146</td>
<td style="text-align: right;">198</td>
<td style="text-align: right;">1.666758</td>
<td style="text-align: right;">0.2415482</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9012346</td>
<td style="text-align: right;">0.9383886</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">132</td>
<td style="text-align: right;">189</td>
<td style="text-align: right;">2.2220139</td>
<td style="text-align: right;">0.0805459</td>
<td style="text-align: right;">0.7568021</td>
<td style="text-align: right;">0.0958904</td>
<td style="text-align: right;">0.0454545</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">28686334-21-T/G</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">161</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">1.309589</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9938272</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">32</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">129</td>
<td style="text-align: right;">156</td>
<td style="text-align: right;">0.7042556</td>
<td style="text-align: right;">0.1754130</td>
<td style="text-align: right;">0.8886749</td>
<td style="text-align: right;">0.1987578</td>
<td style="text-align: right;">0.2606635</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28679782-56-G/T</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">161</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">1.309589</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9938272</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">24</td>
<td style="text-align: right;">144</td>
<td style="text-align: right;">187</td>
<td style="text-align: right;">0.9200279</td>
<td style="text-align: right;">0.8681945</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.1055901</td>
<td style="text-align: right;">0.1137441</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">28685362-20-C/T</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">162</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">1.301530</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">157</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">6.6572377</td>
<td style="text-align: right;">0.0894367</td>
<td style="text-align: right;">0.7891005</td>
<td style="text-align: right;">0.0308642</td>
<td style="text-align: right;">0.0047393</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28688558-63-C/T</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">161</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">1.303397</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9938272</td>
<td style="text-align: right;">0.9952607</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">154</td>
<td style="text-align: right;">203</td>
<td style="text-align: right;">1.3171759</td>
<td style="text-align: right;">0.7845569</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0434783</td>
<td style="text-align: right;">0.0333333</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output consists of a genlight object for each type of loci, plus a results table.</p>
</section>
<section id="run-infer.sex-1" class="level2">
<h2 class="anchored" data-anchor-id="run-infer.sex-1">Run <code>infer.sex</code></h2>
<p>This function uses the output of function filter.sex.linked (list of 6 objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an <code>agreed sex</code>.</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-9_9f947b50c9c9a70600af99165e19c959">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough W-linked/Y-linked loci (need at least 1). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 13%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: left;">#called</th>
<th style="text-align: left;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>What did the function mention about gametologs? How did that impact the results table? How many types of sex-linked loci were used to infer an <code>agreed sex</code>?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>Can you find individuals for which the <code>agreed sex</code> is uncertain (i.e., has an asterisk “*”)?</p>
</div>
</div>
</section>
</section>
<section id="exercise-4" class="level1">
<h1>EXERCISE</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>Imagine you are analyzing the genetic data of a population and you want to look at its genetic structure and genetic diversity. You get to work by filtering your beautiful set of SNPs. Because you are a rigorous, you want to test whether using function <code>filter.sex.linked</code> to properly filter out sex-linked loci changes the results of the populations analyses. So you: (1) process your SNP dataset by applying standard filters and run analyses, and (2) process the dataset by filtering out sex-linked loci first, applying standard filters second, and then run analyses.</p>
<p>Choose one of the following datasets (or your own data) and report:</p>
<ol type="1">
<li><p>How many sex-linked markers are present?</p></li>
<li><p>How many individuals had a wrong sexID?</p></li>
<li><p>Do you see any changes in your PCA/structure analyses when you filtered out sex-linked markers versus when you did not?</p></li>
<li><p>Do you see any differences in genetic diversity and fixation indices when you filtered out sex-linked markers versus when you did not?</p></li>
</ol>
</div>
</div>
<p><img src="images/beer.png" class="class" width="350" height="200"> <img src="images/Eastern_Yellow_Robin.jpg" class="img-fluid" alt="The Eastern Yellow Robin" width="349"></p>
<p><img src="images/Bull_Shark.jpg" class="class" width="350" height="200"> <img src="images/Blue_shark.jpg" class="class" width="350" height="200"></p>
<section id="exercise-data-1---your-own-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-1---your-own-data">Exercise data 1 - Your own data</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
HINT
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can have a look at the exercise data below for inspiration.</p>
</div>
</div>
<section id="number-of-sex-linked-markers" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers">1. Number of sex-linked markers?</h3>
</section>
<section id="individuals-with-wrong-sexid" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid">2. Individuals with wrong sexID?</h3>
</section>
<section id="changes-in-pca-before-and-after-removing-the-slm" class="level3">
<h3 class="anchored" data-anchor-id="changes-in-pca-before-and-after-removing-the-slm">3. Changes in PCA before and after removing the SLM?</h3>
</section>
<section id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" class="level3">
<h3 class="anchored" data-anchor-id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</h3>
</section>
</section>
<section id="exercise-data-2---the-eastern-yellow-robin" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</h2>
<p>Data from Robledo-Ruiz et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Eastern_Yellow_Robin.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Eastern Yellow Robin</figcaption>
</figure>
</div>
<section id="load-data-2" class="level3">
<h3 class="anchored" data-anchor-id="load-data-2">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-10_5ec2bb273e2680f6472c4a6f29a1edd9">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"EYR"</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>EYR<span class="sc">@</span>n.loc</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>pop)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

      F   M 
  1 352 429 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-1" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-1">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-11_b15e8eccdf915ba82a58acff55cb7298">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> EYR, <span class="at">system =</span> <span class="st">"zw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 352 females and 429 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 84 sex-linked loci:
   5 W-linked loci
   54 sex-biased loci
   18 Z-linked loci
   7 ZW gametologs.
And 916 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-1" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-1">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-12_bb2f693a45f33c57344f15a6828b32e1">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>,</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(EYR<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> sexID<span class="sc">$</span>agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: right;">#Hom.g</th>
<th style="text-align: right;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">13</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">15</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">11</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">14</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 80</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>Can you tell which misidentified sexes are due to uncertain genetic sex (indicated with *)?</p>
<p><strong>HINT</strong> Try using <code>grep(pattern = "\\*", x = sexID$agreed.sex)</code></p>
</div>
</div>
</section>
<section id="changes-in-pca-before-and-after-removing-the-slm-1" class="level3">
<h3 class="anchored" data-anchor-id="changes-in-pca-before-and-after-removing-the-slm-1">3. Changes in PCA before and after removing the SLM?</h3>
<section id="processing-snps-with-two-filtering-regimes" class="level4">
<h4 class="anchored" data-anchor-id="processing-snps-with-two-filtering-regimes">Processing SNPs with two filtering regimes</h4>
<section id="filtering-snps-only-with-standard-filters-sloppy" class="level5">
<h5 class="anchored" data-anchor-id="filtering-snps-only-with-standard-filters-sloppy">Filtering SNPs only with standard filters (sloppy)</h5>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-13_9d7998fb0ad0bd0557f67ca748ecfbca">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for read depth</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(EYR)  <span class="co"># This is the initial dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Processing genlight object with SNP data
  Reporting Read Depth by Locus
  No. of loci = 1000 
  No. of individuals = 782 
    Minimum      :  2.6 
    1st quartile :  4.2 
    Median       :  5.5 
    Mean         :  5.9795 
    3r quartile  :  7.5 
    Maximum      :  12.9 
    Missing Rate Overall:  0.19 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%      12.9        1     0.1      999    99.9
2       95%      10.1       51     5.1      949    94.9
3       90%       9.2      103    10.3      897    89.7
4       85%       8.6      151    15.1      849    84.9
5       80%       8.0      204    20.4      796    79.6
6       75%       7.5      254    25.4      746    74.6
7       70%       7.1      302    30.2      698    69.8
8       65%       6.6      354    35.4      646    64.6
9       60%       6.2      405    40.5      595    59.5
10      55%       5.9      452    45.2      548    54.8
11      50%       5.5      512    51.2      488    48.8
12      45%       5.2      566    56.6      434    43.4
13      40%       4.9      616    61.6      384    38.4
14      35%       4.7      661    66.1      339    33.9
15      30%       4.4      718    71.8      282    28.2
16      25%       4.2      759    75.9      241    24.1
17      20%       4.0      808    80.8      192    19.2
18      15%       3.8      851    85.1      149    14.9
19      10%       3.5      903    90.3       97     9.7
20       5%       3.2      952    95.2       48     4.8
21       0%       2.6     1000   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(EYR, <span class="at">lower =</span> <span class="dv">3</span>, <span class="at">upper =</span> <span class="dv">11</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for loci call rate</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"loc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data
  Reporting Call Rate by Locus
  No. of loci = 957 
  No. of individuals = 782 
    Minimum      :  0.213555 
    1st quartile :  0.726343 
    Median       :  0.886189 
    Mean         :  0.8147825 
    3r quartile  :  0.950128 
    Maximum      :  0.98977 
    Missing Rate Overall:  0.1852 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"loc"</span>,</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="fl">0.75</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for individual call rate</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data

  Reporting Call Rate by Individual
  No. of loci = 694 
  No. of individuals = 782 
    Minimum      :  0.03746398 
    1st quartile :  0.9135447 
    Median       :  0.9394813 
    Mean         :  0.9094375 
    3r quartile  :  0.9538905 
    Maximum      :  0.9783862 
    Missing Rate Overall:  0.0906 

Listing 4 populations and their average CallRates
  Monitor again after filtering
  Population CallRate   N
1     Crusoe   0.9017 238
2 Muckleford   0.9058 421
3      Timor   0.9411  52
4     Wombat   0.9338  71

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual   CallRate
1    M18.29.1 0.03746398
2    M18.18.1 0.04034582
3    M18.47.2 0.07204611
4    C18.21.2 0.09654179
5   027-34168 0.09942363
6    C18.15.2 0.09942363
7    C18.16.1 0.10951009
8    M18.47.3 0.12103746
9    M18.35.2 0.15417867
10   M18.20.3 0.27521614
11   C18.28.1 0.42939481
12   M20.70.2 0.43948127
13   C18.17.2 0.45244957
14  027-34065 0.48847262
15   C18.14.1 0.49711816
16   M20.70.3 0.50864553
17  M20.110.1 0.51585014
18    M19.8.1 0.53314121
19   M19.12.1 0.57060519
20   M20.64.3 0.58213256

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"ind"</span>,</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="fl">0.65</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for MAC (= 3)</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(EYR.sloppy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
  Processing genlight object with SNP data
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Crusoe 
  No. of loci = 668 
  No. of individuals = 231 
    Minimum      :  0.0022 
    1st quantile :  0.066475 
    Median       :  0.147 
    Mean         :  0.1822217 
    3r quantile  :  0.282275 
    Maximum      :  0.5 
    Missing Rate Overall:  0.08 

  Reporting Minor Allele Frequency (MAF) by Locus for population Muckleford 
  No. of loci = 674 
  No. of individuals = 406 
    Minimum      :  0.0012 
    1st quantile :  0.059975 
    Median       :  0.1432 
    Mean         :  0.1777687 
    3r quantile  :  0.2759 
    Maximum      :  0.4987 
    Missing Rate Overall:  0.07 

  Reporting Minor Allele Frequency (MAF) by Locus for population Timor 
  No. of loci = 588 
  No. of individuals = 52 
    Minimum      :  0.0096 
    1st quantile :  0.0673 
    Median       :  0.1651 
    Mean         :  0.1945571 
    3r quantile  :  0.30445 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus for population Wombat 
  No. of loci = 626 
  No. of individuals = 71 
    Minimum      :  0.007 
    1st quantile :  0.064525 
    Median       :  0.14685 
    Mean         :  0.1810583 
    3r quantile  :  0.2752 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 694 
  No. of individuals = 760 
    Minimum      :  3e-04 
    1st quantile :  0.061925 
    Median       :  0.139 
    Mean         :  0.1744435 
    3r quantile  :  0.272375 
    Maximum      :  0.4954 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.4954        1     0.1      689    99.9
2       95%    0.4444       35     5.1      655    94.9
3       90%    0.3799       70    10.1      620    89.9
4       85%    0.3485      104    15.1      586    84.9
5       80%    0.3091      139    20.1      551    79.9
6       75%    0.2728      173    25.1      517    74.9
7       70%    0.2368      207    30.0      483    70.0
8       65%    0.2100      242    35.1      448    64.9
9       60%    0.1892      276    40.0      414    60.0
10      55%    0.1585      311    45.1      379    54.9
11      50%    0.1387      346    50.1      344    49.9
12      45%    0.1218      381    55.2      309    44.8
13      40%    0.1027      416    60.3      274    39.7
14      35%    0.0868      449    65.1      241    34.9
15      30%    0.0729      483    70.0      207    30.0
16      25%    0.0619      518    75.1      172    24.9
17      20%    0.0500      553    80.1      137    19.9
18      15%    0.0358      587    85.1      103    14.9
19      10%    0.0221      622    90.1       68     9.9
20       5%    0.0077      656    95.1       34     4.9
21       0%    0.0003      690   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(EYR.sloppy, <span class="at">threshold =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="dv">0</span>,</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.select.colors 
  Warning: Number of required colors not specified, set to 9
  Library: RColorBrewer
  Palette: brewer.pal
  Showing and returning 2 of 9 colors for library RColorBrewer : palette Blues </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.select.colors </code></pre>
</div>
</div>
</section>
<section id="filtering-snps-with-filter.sex.linked-and-standard-filters-correct" class="level5">
<h5 class="anchored" data-anchor-id="filtering-snps-with-filter.sex.linked-and-standard-filters-correct">Filtering SNPs with filter.sex.linked and standard filters (correct)</h5>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-14_749e415d9badd944f4b78c0c49fef07a">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for sex-linked loci</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>correct <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(EYR, <span class="at">system =</span> <span class="st">"zw"</span>)  <span class="co"># This is the initial dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-4.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will use correct$autosomal for the next filters</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for read depth</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(correct<span class="sc">$</span>autosomal)  <span class="co"># This is the filtered dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Processing genlight object with SNP data
  Reporting Read Depth by Locus
  No. of loci = 916 
  No. of individuals = 782 
    Minimum      :  2.6 
    1st quartile :  4.2 
    Median       :  5.6 
    Mean         :  6.023253 
    3r quartile  :  7.5 
    Maximum      :  12.9 
    Missing Rate Overall:  0.18 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%      12.9        1     0.1      915    99.9
2       95%      10.1       50     5.5      866    94.5
3       90%       9.3       93    10.2      823    89.8
4       85%       8.7      141    15.4      775    84.6
5       80%       8.1      185    20.2      731    79.8
6       75%       7.5      241    26.3      675    73.7
7       70%       7.1      285    31.1      631    68.9
8       65%       6.7      329    35.9      587    64.1
9       60%       6.2      379    41.4      537    58.6
10      55%       6.0      413    45.1      503    54.9
11      50%       5.6      465    50.8      451    49.2
12      45%       5.3      510    55.7      406    44.3
13      40%       5.0      551    60.2      365    39.8
14      35%       4.7      604    65.9      312    34.1
15      30%       4.4      657    71.7      259    28.3
16      25%       4.2      694    75.8      222    24.2
17      20%       4.0      738    80.6      178    19.4
18      15%       3.8      779    85.0      137    15.0
19      10%       3.5      827    90.3       89     9.7
20       5%       3.2      872    95.2       44     4.8
21       0%       2.6      916   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(correct<span class="sc">$</span>autosomal, <span class="at">lower =</span> <span class="dv">3</span>,</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">upper =</span> <span class="dv">11</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-6.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for loci call rate</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"loc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data
  Reporting Call Rate by Locus
  No. of loci = 874 
  No. of individuals = 782 
    Minimum      :  0.213555 
    1st quartile :  0.7317778 
    Median       :  0.890026 
    Mean         :  0.8179257 
    3r quartile  :  0.9523655 
    Maximum      :  0.98977 
    Missing Rate Overall:  0.1821 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"loc"</span>,</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="fl">0.75</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for individual call rate</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data

  Reporting Call Rate by Individual
  No. of loci = 639 
  No. of individuals = 782 
    Minimum      :  0.04068858 
    1st quartile :  0.9170579 
    Median       :  0.9405321 
    Mean         :  0.9114205 
    3r quartile  :  0.9557903 
    Maximum      :  0.9780908 
    Missing Rate Overall:  0.0886 

Listing 4 populations and their average CallRates
  Monitor again after filtering
  Population CallRate   N
1     Crusoe   0.9028 238
2 Muckleford   0.9085 421
3      Timor   0.9428  52
4     Wombat   0.9347  71

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual   CallRate
1    M18.29.1 0.04068858
2    M18.18.1 0.04381847
3    M18.47.2 0.07198748
4    C18.15.2 0.09702660
5    C18.21.2 0.09859155
6   027-34168 0.10172144
7    C18.16.1 0.10798122
8    M18.47.3 0.12206573
9    M18.35.2 0.15492958
10   M18.20.3 0.28638498
11   C18.28.1 0.44131455
12   M20.70.2 0.44757433
13   C18.17.2 0.44913928
14  027-34065 0.48513302
15   C18.14.1 0.49921753
16  M20.110.1 0.51956182
17   M20.70.3 0.51956182
18    M19.8.1 0.54303599
19   M19.12.1 0.57589984
20   M20.64.3 0.59780908

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"ind"</span>,</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="fl">0.65</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for MAC (= 3)</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(EYR.correct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
  Processing genlight object with SNP data
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Crusoe 
  No. of loci = 613 
  No. of individuals = 231 
    Minimum      :  0.0022 
    1st quantile :  0.0667 
    Median       :  0.145 
    Mean         :  0.1795251 
    3r quantile  :  0.2814 
    Maximum      :  0.5 
    Missing Rate Overall:  0.08 

  Reporting Minor Allele Frequency (MAF) by Locus for population Muckleford 
  No. of loci = 620 
  No. of individuals = 406 
    Minimum      :  0.0012 
    1st quantile :  0.05805 
    Median       :  0.1341 
    Mean         :  0.1736115 
    3r quantile  :  0.272125 
    Maximum      :  0.4987 
    Missing Rate Overall:  0.07 

  Reporting Minor Allele Frequency (MAF) by Locus for population Timor 
  No. of loci = 540 
  No. of individuals = 52 
    Minimum      :  0.0096 
    1st quantile :  0.066325 
    Median       :  0.15825 
    Mean         :  0.1911578 
    3r quantile  :  0.296325 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus for population Wombat 
  No. of loci = 572 
  No. of individuals = 71 
    Minimum      :  0.007 
    1st quantile :  0.063175 
    Median       :  0.1449 
    Mean         :  0.1799673 
    3r quantile  :  0.2754 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 639 
  No. of individuals = 760 
    Minimum      :  3e-04 
    1st quantile :  0.0575 
    Median       :  0.1353 
    Mean         :  0.1702414 
    3r quantile  :  0.26845 
    Maximum      :  0.4954 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.4954        1     0.2      634    99.8
2       95%    0.4397       32     5.0      603    95.0
3       90%    0.3738       64    10.1      571    89.9
4       85%    0.3457       96    15.1      539    84.9
5       80%    0.3006      128    20.2      507    79.8
6       75%    0.2686      159    25.0      476    75.0
7       70%    0.2298      191    30.1      444    69.9
8       65%    0.2021      223    35.1      412    64.9
9       60%    0.1779      254    40.0      381    60.0
10      55%    0.1513      286    45.0      349    55.0
11      50%    0.1353      318    50.1      317    49.9
12      45%    0.1141      350    55.1      285    44.9
13      40%    0.0984      383    60.3      252    39.7
14      35%    0.0823      413    65.0      222    35.0
15      30%    0.0689      446    70.2      189    29.8
16      25%    0.0573      477    75.1      158    24.9
17      20%    0.0478      509    80.2      126    19.8
18      15%    0.0348      540    85.0       95    15.0
19      10%    0.0216      572    90.1       63     9.9
20       5%    0.0073      604    95.1       31     4.9
21       0%    0.0003      635   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(EYR.correct, <span class="at">threshold =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="dv">0</span>,</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.select.colors 
  Warning: Number of required colors not specified, set to 9
  Library: RColorBrewer
  Palette: brewer.pal
  Showing and returning 2 of 9 colors for library RColorBrewer : palette Blues </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.select.colors </code></pre>
</div>
</div>
</section>
</section>
<section id="pca-on-sloppy-dataset-only-standard-filters" class="level4">
<h4 class="anchored" data-anchor-id="pca-on-sloppy-dataset-only-standard-filters">PCA on sloppy dataset (only standard filters)</h4>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-15_fd979c6f7f1ca9aa45b194854adb0392">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>PCA.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.sloppy, EYR.sloppy, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.sloppy, EYR.sloppy, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
<section id="pca-on-correct-dataset-filter.sex.linked-and-standard-filters" class="level4">
<h4 class="anchored" data-anchor-id="pca-on-correct-dataset-filter.sex.linked-and-standard-filters">PCA on correct dataset (filter.sex.linked and standard filters)</h4>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-16_087336cd7b83017d1d901d8a78ca29ab">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>PCA.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.correct, EYR.correct, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.correct, EYR.correct, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
</section>
<section id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" class="level3">
<h3 class="anchored" data-anchor-id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-17_a23c402ab2ad949573c240c8c622431b">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic stats</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>basic.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(EYR.sloppy)</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>basic.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(EYR.correct)</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>basic.sloppy<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.1632  0.2417  0.2494  0.0078  0.2518  0.0136  0.0311  0.0538  0.3247  0.0238 
Gst_max   Gst_H 
 0.7099  0.0758 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>basic.correct<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.1656  0.2400  0.2467  0.0067  0.2491  0.0121  0.0272  0.0487  0.3098  0.0213 
Gst_max   Gst_H 
 0.7119  0.0684 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Genetic diversity per pop</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>divers.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(EYR.sloppy, <span class="at">pbar =</span> <span class="cn">FALSE</span>,</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">table =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>divers.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(EYR.correct, <span class="at">pbar =</span> <span class="cn">FALSE</span>,</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">table =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>divers.sloppy<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.3899117  0.3882628  0.3600321  0.3662601 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>divers.correct<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.3887673  0.3844011  0.3574010  0.3646236 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>divers.sloppy<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Crusoe Muckleford      Timor     Wombat
Crusoe              NA 0.02591592 0.08428179 0.05744001
Muckleford 0.006242194         NA 0.06589556 0.06332248
Timor      0.020357562 0.02118200         NA 0.08530756
Wombat     0.017243564 0.01806800 0.03218337         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>divers.correct<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Crusoe Muckleford      Timor     Wombat
Crusoe              NA 0.02470901 0.08101848 0.05574828
Muckleford 0.005281376         NA 0.06535853 0.06220802
Timor      0.018781429 0.02096451         NA 0.08157260
Wombat     0.015170153 0.01735323 0.03085328         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fixation indices</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.02647926         NA         NA     NA
Timor      0.04091224 0.04504432         NA     NA
Wombat     0.05056885 0.02243625 0.07085069     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.02392518         NA         NA     NA
Timor      0.04070621 0.04145570         NA     NA
Wombat     0.04059252 0.01890841 0.06063922     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Your plot was not shown in full because your 'Plots' pane
    is too small. Increase the size of the 'Plots' pane and run the
    function again. Alternatively, use the parameter 'plot.file' to
    save the plot to a file.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0122 0.0173 0.0232
Muckleford 0.0122         NA 0.0199 0.0086
Timor      0.0173     0.0199     NA 0.0312
Wombat     0.0232     0.0086 0.0312     NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0304 0.0549 0.0622
Muckleford 0.0304         NA 0.0583 0.0319
Timor      0.0549     0.0583     NA 0.0900
Wombat     0.0622     0.0319 0.0900     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0204 0.0361 0.0411
Muckleford 0.0204         NA 0.0382 0.0206
Timor      0.0361     0.0382     NA 0.0578
Wombat     0.0411     0.0206 0.0578     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0485 0.0862 0.0980
Muckleford 0.0485         NA 0.0914 0.0501
Timor      0.0862     0.0914     NA 0.1392
Wombat     0.0980     0.0501 0.1392     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0122                      0.0173
Fstp                            0.0304                      0.0549
Dest                            0.0204                      0.0361
Gst_H                           0.0485                      0.0862
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0232                          0.0199
Fstp                        0.0622                          0.0583
Dest                        0.0411                          0.0382
Gst_H                       0.0980                          0.0914
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0086                      0.0312
Fstp                            0.0319                      0.0900
Dest                            0.0206                      0.0578
Gst_H                           0.0501                      0.1392</code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Your plot was not shown in full because your 'Plots' pane
    is too small. Increase the size of the 'Plots' pane and run the
    function again. Alternatively, use the parameter 'plot.file' to
    save the plot to a file.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0109 0.0173 0.0179
Muckleford 0.0109         NA 0.0180 0.0068
Timor      0.0173     0.0180     NA 0.0258
Wombat     0.0179     0.0068 0.0258     NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0278 0.0548 0.0517
Muckleford 0.0278         NA 0.0546 0.0282
Timor      0.0548     0.0546     NA 0.0793
Wombat     0.0517     0.0282 0.0793     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0184 0.0357 0.0338
Muckleford 0.0184         NA 0.0351 0.0180
Timor      0.0357     0.0351     NA 0.0503
Wombat     0.0338     0.0180 0.0503     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0441 0.0858 0.0812
Muckleford 0.0441         NA 0.0850 0.0440
Timor      0.0858     0.0850     NA 0.1224
Wombat     0.0812     0.0440 0.1224     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0109                      0.0173
Fstp                            0.0278                      0.0548
Dest                            0.0184                      0.0357
Gst_H                           0.0441                      0.0858
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0179                          0.0180
Fstp                        0.0517                          0.0546
Dest                        0.0338                          0.0351
Gst_H                       0.0812                          0.0850
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0068                      0.0258
Fstp                            0.0282                      0.0793
Dest                            0.0180                      0.0503
Gst_H                           0.0440                      0.1224</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-3---bull-shark" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-3---bull-shark">Exercise data 3 - Bull shark</h2>
<p>Data from Devloo-Delva et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Bull_Shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Bull Shark</figcaption>
</figure>
</div>
<section id="load-data-3" class="level3">
<h3 class="anchored" data-anchor-id="load-data-3">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-18_666cb9b3849324c9a34b573b8d33cd66">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Bull_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 93202</code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
  85   64   33 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-2" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-2">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-19_e32626452543f0c8b761afa4a43f0f06">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 85 females and 64 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 93202.
Found 9 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   4 X-linked loci
   1 XY gametologs.
And 93193 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-2" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-2">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-20_0e28f2482d7687bfc477d3465cfdd290">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex)  <span class="co"># remove asterisk</span></span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">60088</td>
<td style="text-align: left;">60088</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60160</td>
<td style="text-align: left;">60160</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60168</td>
<td style="text-align: left;">60168</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60176</td>
<td style="text-align: left;">60176</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60096</td>
<td style="text-align: left;">60096</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60104</td>
<td style="text-align: left;">60104</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 22</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-4---blue-shark" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-4---blue-shark">Exercise data 4 - Blue shark</h2>
<p>Data from Nikolic et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Blue_shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Blue Shark</figcaption>
</figure>
</div>
<section id="load-data-4" class="level3">
<h3 class="anchored" data-anchor-id="load-data-4">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-21_7d73ed2a5470ae634b246ff419204a03">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Blue_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 172384</code></pre>
</div>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
 104  111  149 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-3" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-3">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-22_1d9d1c818893e98334aa75aedcdcabf4">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 104 females and 111 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-22-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 172384.
Found 26 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   22 X-linked loci
   0 XY gametologs.
And 172358 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-22-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-3" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-3">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-23_82203f83fb48163c328a127548cfcf51">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex)  <span class="co"># remove asterisk</span></span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">60088</td>
<td style="text-align: left;">60088</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60160</td>
<td style="text-align: left;">60160</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60168</td>
<td style="text-align: left;">60168</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60176</td>
<td style="text-align: left;">60176</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60096</td>
<td style="text-align: left;">60096</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60104</td>
<td style="text-align: left;">60104</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 22</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Devloo-Delva, F., Burridge, C. P., Kyne, P. M., Brunnschweiler, J. M., Chapman, D. D., Charvet, P., … &amp; Feutry, P. (2023). From rivers to ocean basins: The role of ocean barriers and philopatry in the genetic structuring of a cosmopolitan coastal predator. Ecology and Evolution, 13(2), e9837. https://doi.org/10.1002/ece3.9837</p>
<p>Nikolic, N., Devloo-Delva, F., Bailleul, D., Noskova, E., Rougeux, C., Delord, C., … &amp; Arnaud‐Haond, S. (2023). Stepping up to genome scan allows stock differentiation in the worldwide distributed blue shark <em>Prionace glauca</em>. Molecular Ecology, 32(5), 1000-1019. https://doi.org/10.1111/mec.16822</p>
<p>Robledo-Ruiz, D. A., Austin, L., Amos, J. N., Castrejón-Figueroa, J., Harley, D. K. P., Magrath, M. J. L., Sunnucks, P., &amp; Pavlova, A. (2023). Easy-to-use R functions to separate reduced-representation genomic datasets into sex-linked and autosomal loci, and conduct sex assignment. Molecular Ecology Resources, 00, 1–21. https://doi.org/10.1111/1755-0998.13844</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>