---
title: "Session 10: Sex-linked markers"
authors: "Diana Robledo-Ruiz & Floriaan Devloo-Delva"
format: 
  html:
    theme: cosmo
    css: "css.css"
    keep-tex: true
    toc: true
    toc_depth: 3
    toc_float: yes
    toc-location: left
    number-sections: false
editor: visual
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(knitr)
library(formatR)
library(tidyverse)
knitr::opts_chunk$set(tidy = TRUE, tidy.opts = list(width.cutoff = 72), 
                      cache = TRUE, 
                      echo = TRUE)
```

# Required packages

```{r, warning=FALSE, message=FALSE}
library(dartR.base)
library(dartR.sexlinked)
```


# Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater

![The Yellow Tufted Honeyeater](images/Yellow_Tufted_Honeyeater.jpg)

## Load data

```{r}
data("YTH")
```

## Run `filter.sex.linked`
This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.

*The genlight object must contain in gl@other$ind.metrics a column named "id", and a column named "sex" in which individuals with known-sex are assigned 'M' for male, or 'F' for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).*

![ZW/ZZ sex chromosomes](images/ZW.png)

::: {.infobox .task data-latex="{task}"}
**NOTE**

Set `ncores = 1` if you have less than 50,000 SNPs, since it could actually slow down the analysis.
:::

```{r, results="hold"}
res <- dartR.sexlinked::filter.sex.linked(gl = YTH, system = "zw", plots = TRUE, ncores = 1)
```


## Run `infer.sex`

This function uses the output of function filter.sex.linked (list of 6 genlight objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an agreed sex.

```{r}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "zw", seed = 124)

knitr::kable(head(sexID))
```

::: {.infobox .task data-latex="{task}"}
**IMPORTANT**

Any `agreed sex` with an asterisk* need to manually checked.
:::

# Dataset 2 - XX/XY - The Leadbeater's possum

![The Leadbeater's possum](images/Leadbeaters_possum.jpg)

## Load data

```{r}
data("LBP")
```

## Run `filter.sex.linked`

This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.

*The genlight object must contain in gl@other$ind.metrics a column named "id", and a column named "sex" in which individuals with known-sex are assigned 'M' for male, or 'F' for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).*

![XX/XY sex chromosomes](images/XY.png)

```{r, results="hold"}
res <- dartR.sexlinked::filter.sex.linked(gl = LBP, system = "xy", plots = TRUE, ncores = 1)
```

## Run `infer.sex`

This function uses the output of function filter.sex.linked (list of 6 genlight objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an agreed sex.

```{r}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "xy", seed = 124)
knitr::kable(head(sexID))
```


# EXCERCISE

::: callout-note
## Exercise

![](images/task.png){#id .class width="48" height="48"} 

Choose one of the following dataset (or your own data) and report:

1.  How many sex-linked markers are present?

2.  How many individuals had a wrong sexID?

3.  Do you see any changes in your PCA/structure analyses before and after filtering the sex-linked markers?

4.  Do you see any differences in genetic diversity and fixation indices between autosomal and sex-linked markers?
:::

![](images/beer.png){#id .class width="350" height="200"} ![](images/Leadbeaters_possum.jpg){#id .class width="350" height="200"}
![](images/Bull_Shark.jpg){#id .class width="350" height="200"} ![](images/Blue_shark.jpg){#id .class width="350" height="200"}

## Exercise data 1 - Your own data

::: {.infobox .task data-latex="{task}"}
**HINT**

You can have a look at the exercise data below for inspiration.
:::

### 1. Number of sex-linked markers?
### 2. Individuals with wrong sexID?
### 3. Changes in PCA before and after removing the SLM?
### 4. Differences in genetic diversity and fixation indices between autosomal and SLM?


## Exercise data 2 - The Eastern Yellow Robin

Data from Robledo-Ruiz et al. 2023.


![The Eastern Yellow Robin](images/Eastern_Yellow_Robin.jpg)

### Load data

```{r, results="hold"}
data("EYR")

EYR@n.loc
table(EYR@pop)
table(EYR@other$ind.metrics$pop)
table(EYR@other$ind.metrics$sex, useNA = "ifany")
```

### 1. Number of sex-linked markers?

```{r, results="hold"}
res <- dartR.sexlinked::filter.sex.linked(gl = EYR, system = "zw", plots = TRUE, ncores = 1)
```

### 2. Individuals with wrong sexID?

```{r, results="hold"}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "zw", seed = 124)
knitr::kable(head(sexID))

agreed.sex <- sub(pattern = "\\*", replacement = "", x = sexID$agreed.sex) # remove asterisk
sum(EYR$other$ind.metrics$sex != agreed.sex, na.rm = TRUE)
```

::: callout-note
#### Exercise

![](images/task.png){#id .class width="48" height="48"} 

Can you check if any misidentified sexes are due to uncertain genetic sexID (indicated with *)?

**HINT**
Try using `grep(pattern = "\\*", x = sexID$agreed.sex)`
:::

### 3. Changes in PCA before and after removing the SLM?

#### PCA before

```{r, results="hold", warning=FALSE}
PCA.before <- dartR.base::gl.pcoa(EYR, verbose = 0)
dartR.base::gl.pcoa.plot(PCA.before, EYR, xaxis = 1, yaxis = 2)
dartR.base::gl.pcoa.plot(PCA.before, EYR, xaxis = 2, yaxis = 3)
```

#### Filtering

```{r, message=FALSE, warning=FALSE}
EYR.after <- res$autosomal
dartR.base::gl.report.rdepth(EYR.after)
EYR.after <- dartR.base::gl.filter.rdepth(EYR.after, lower = 3, upper = 11, verbose = 0)
dartR.base::gl.report.callrate(EYR.after, method = "loc")
EYR.after <- dartR.base::gl.filter.callrate(EYR, method = "loc",  threshold = 0.75, verbose = 0)
dartR.base::gl.report.callrate(EYR, method = "ind")
EYR.after <- dartR.base::gl.filter.callrate(EYR.after, method = "ind", threshold = 0.65, verbose = 0)
dartR.base::gl.report.maf(EYR.after)
EYR.after <- dartR.base::gl.filter.maf(EYR.after, threshold = 0.05, verbose = 0)
```

#### PCA after

```{r, results="hold", warning=FALSE}
PCA.after <- dartR.base::gl.pcoa(EYR.after, verbose = 0)
dartR.base::gl.pcoa.plot(PCA.after, EYR.after, xaxis = 1, yaxis = 2)
dartR.base::gl.pcoa.plot(PCA.after, EYR.after, xaxis = 2, yaxis = 3)
```

### 4. Differences in genetic diversity and fixation indices between autosomal and SLM?

```{r, warning=FALSE}
autosomal <- res$autosomal
z.linked <- res$z.linked
z.linked.F <- z.linked[z.linked$other$ind.metrics$sex == "F",]
z.linked.M <- z.linked[z.linked$other$ind.metrics$sex == "M",]

# Genetic diversity

basic.autosomal <- dartR.base::utils.basic.stats(autosomal)
basic.autosomal$overall
basic.xF <- dartR.base::utils.basic.stats(z.linked.F)
basic.xF$overall
basic.xM <- dartR.base::utils.basic.stats(z.linked.M)
basic.xM$overall

divers.auto <- dartR.base::gl.report.diversity(autosomal, pbar = FALSE, table = FALSE, verbose = 0)
divers.auto$one_H_alpha
divers.auto$one_H_beta
divers.zF <- dartR.base::gl.report.diversity(z.linked.F, pbar = FALSE, table = FALSE, verbose = 0)
divers.zF$one_H_alpha
divers.zF$one_H_beta
divers.zM <- dartR.base::gl.report.diversity(z.linked.M, pbar = FALSE, table = FALSE, verbose = 0)
divers.zM$one_H_alpha
divers.zM$one_H_beta

# Fixation indices
dartR.base::gl.fst.pop(autosomal, verbose = 0)
dartR.base::gl.fst.pop(z.linked.F, verbose = 0)
dartR.base::gl.fst.pop(z.linked.M, verbose = 0)

dartR.base::gl.report.fstat(autosomal, verbose = 0)
dartR.base::gl.report.fstat(z.linked.F, verbose = 0)
dartR.base::gl.report.fstat(z.linked.F, verbose = 0)
```

## Exercise data 3 - Bull shark

Data from Devloo-Delva et al. 2023.

![The Bull Shark](images/Bull_Shark.jpg)

### Load data

```{r}
print(load("data/Bull_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"))

data.gl@n.loc
table(data.gl@pop)
table(data.gl@other$ind.metrics$pop)
table(data.gl@other$ind.metrics$sex, useNA = "ifany")
```

### 1. Number of sex-linked markers?

```{r, results="hold"}
res <- dartR.sexlinked::filter.sex.linked(gl = data.gl, system = "xy", plots = TRUE, ncores = 1)
```

### 2. Individuals with wrong sexID?

```{r, results="hold"}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "xy", seed = 124)
knitr::kable(head(sexID))

agreed.sex <- sub(pattern = "\\*", replacement = "", x = sexID$agreed.sex) # remove asterisk
sum(data.gl$other$ind.metrics$sex != agreed.sex, na.rm = TRUE)
```

## Exercise data 4 - Blue shark

Data from Nikolic et al. 2023.

![The Blue Shark](images/Blue_shark.jpg)

### Load data

```{r}
print(load("data/Blue_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"))

data.gl@n.loc
table(data.gl@pop)
table(data.gl@other$ind.metrics$pop)
table(data.gl@other$ind.metrics$sex, useNA = "ifany")
```

### 1. Number of sex-linked markers?

```{r, results="hold"}
res <- dartR.sexlinked::filter.sex.linked(gl = data.gl, system = "xy", plots = TRUE, ncores = 1)
```

### 2. Individuals with wrong sexID?

```{r, results="hold"}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "xy", seed = 124)
knitr::kable(head(sexID))

agreed.sex <- sub(pattern = "\\*", replacement = "", x = sexID$agreed.sex) # remove asterisk
sum(data.gl$other$ind.metrics$sex != agreed.sex, na.rm = TRUE)
```


# References

Devloo-Delva, F., Burridge, C. P., Kyne, P. M., Brunnschweiler, J. M., Chapman, D. D., Charvet, P., ... & Feutry, P. (2023). From rivers to ocean basins: The role of ocean barriers and philopatry in the genetic structuring of a cosmopolitan coastal predator. Ecology and Evolution, 13(2), e9837. https://doi.org/10.1002/ece3.9837

Nikolic, N., Devloo-Delva, F., Bailleul, D., Noskova, E., Rougeux, C., Delord, C., ... & Arnaud‐Haond, S. (2023). Stepping up to genome scan allows stock differentiation in the worldwide distributed blue shark *Prionace glauca*. Molecular Ecology, 32(5), 1000-1019. https://doi.org/10.1111/mec.16822

Robledo-Ruiz, D. A., Austin, L., Amos, J. N., Castrejón-Figueroa, J., Harley, D. K. P., Magrath, M. J. L., Sunnucks, P., & Pavlova, A. (2023). Easy-to-use R functions to separate reduced-representation genomic datasets into sex-linked and autosomal loci, and conduct sex assignment. Molecular Ecology Resources, 00, 1–21. https://doi.org/10.1111/1755-0998.13844
