<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Session 10: Sex-linked markers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Session10_SexLinkedMarkers_files/libs/clipboard/clipboard.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/quarto.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/popper.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Session10_SexLinkedMarkers_files/libs/quarto-html/anchor.min.js"></script>
<link href="Session10_SexLinkedMarkers_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Session10_SexLinkedMarkers_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Session10_SexLinkedMarkers_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


<link rel="stylesheet" href="css.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages">Required packages</a></li>
  <li><a href="#dataset-1---zwzz---the-yellow-tufted-honeyeater" id="toc-dataset-1---zwzz---the-yellow-tufted-honeyeater" class="nav-link" data-scroll-target="#dataset-1---zwzz---the-yellow-tufted-honeyeater">Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#run-filter.sex.linked" id="toc-run-filter.sex.linked" class="nav-link" data-scroll-target="#run-filter.sex.linked">Run <code>filter.sex.linked</code></a></li>
  <li><a href="#run-infer.sex" id="toc-run-infer.sex" class="nav-link" data-scroll-target="#run-infer.sex">Run <code>infer.sex</code></a></li>
  </ul></li>
  <li><a href="#dataset-2---xxxy---the-leadbeaters-possum" id="toc-dataset-2---xxxy---the-leadbeaters-possum" class="nav-link" data-scroll-target="#dataset-2---xxxy---the-leadbeaters-possum">Dataset 2 - XX/XY - The Leadbeater’s possum</a>
  <ul class="collapse">
  <li><a href="#load-data-1" id="toc-load-data-1" class="nav-link" data-scroll-target="#load-data-1">Load data</a></li>
  <li><a href="#run-filter.sex.linked-1" id="toc-run-filter.sex.linked-1" class="nav-link" data-scroll-target="#run-filter.sex.linked-1">Run <code>filter.sex.linked</code></a></li>
  <li><a href="#run-infer.sex-1" id="toc-run-infer.sex-1" class="nav-link" data-scroll-target="#run-infer.sex-1">Run <code>infer.sex</code></a></li>
  </ul></li>
  <li><a href="#excercise" id="toc-excercise" class="nav-link" data-scroll-target="#excercise">EXCERCISE</a>
  <ul class="collapse">
  <li><a href="#exercise-data-1---your-own-data" id="toc-exercise-data-1---your-own-data" class="nav-link" data-scroll-target="#exercise-data-1---your-own-data">Exercise data 1 - Your own data</a>
  <ul class="collapse">
  <li><a href="#number-of-sex-linked-markers" id="toc-number-of-sex-linked-markers" class="nav-link" data-scroll-target="#number-of-sex-linked-markers">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid" id="toc-individuals-with-wrong-sexid" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid">2. Individuals with wrong sexID?</a></li>
  <li><a href="#changes-in-pca-before-and-after-removing-the-slm" id="toc-changes-in-pca-before-and-after-removing-the-slm" class="nav-link" data-scroll-target="#changes-in-pca-before-and-after-removing-the-slm">3. Changes in PCA before and after removing the SLM?</a></li>
  <li><a href="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" id="toc-differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" class="nav-link" data-scroll-target="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</a></li>
  </ul></li>
  <li><a href="#exercise-data-2---the-eastern-yellow-robin" id="toc-exercise-data-2---the-eastern-yellow-robin" class="nav-link" data-scroll-target="#exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</a>
  <ul class="collapse">
  <li><a href="#load-data-2" id="toc-load-data-2" class="nav-link" data-scroll-target="#load-data-2">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-1" id="toc-number-of-sex-linked-markers-1" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-1">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-1" id="toc-individuals-with-wrong-sexid-1" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-1">2. Individuals with wrong sexID?</a></li>
  <li><a href="#changes-in-pca-before-and-after-removing-the-slm-1" id="toc-changes-in-pca-before-and-after-removing-the-slm-1" class="nav-link" data-scroll-target="#changes-in-pca-before-and-after-removing-the-slm-1">3. Changes in PCA before and after removing the SLM?</a></li>
  <li><a href="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" id="toc-differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" class="nav-link" data-scroll-target="#differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</a></li>
  </ul></li>
  <li><a href="#exercise-data-3---bull-shark" id="toc-exercise-data-3---bull-shark" class="nav-link" data-scroll-target="#exercise-data-3---bull-shark">Exercise data 3 - Bull shark</a>
  <ul class="collapse">
  <li><a href="#load-data-3" id="toc-load-data-3" class="nav-link" data-scroll-target="#load-data-3">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-2" id="toc-number-of-sex-linked-markers-2" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-2">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-2" id="toc-individuals-with-wrong-sexid-2" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-2">2. Individuals with wrong sexID?</a></li>
  </ul></li>
  <li><a href="#exercise-data-4---blue-shark" id="toc-exercise-data-4---blue-shark" class="nav-link" data-scroll-target="#exercise-data-4---blue-shark">Exercise data 4 - Blue shark</a>
  <ul class="collapse">
  <li><a href="#load-data-4" id="toc-load-data-4" class="nav-link" data-scroll-target="#load-data-4">Load data</a></li>
  <li><a href="#number-of-sex-linked-markers-3" id="toc-number-of-sex-linked-markers-3" class="nav-link" data-scroll-target="#number-of-sex-linked-markers-3">1. Number of sex-linked markers?</a></li>
  <li><a href="#individuals-with-wrong-sexid-3" id="toc-individuals-with-wrong-sexid-3" class="nav-link" data-scroll-target="#individuals-with-wrong-sexid-3">2. Individuals with wrong sexID?</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Session 10: Sex-linked markers</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Diana Robledo-Ruiz &amp; Floriaan Devloo-Delva </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="required-packages" class="level1">
<h1>Required packages</h1>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-1_1c98dd5edc3b6ca21602712ef92c5e0c">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.base)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.sexlinked)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dataset-1---zwzz---the-yellow-tufted-honeyeater" class="level1">
<h1>Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Yellow_Tufted_Honeyeater.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Yellow Tufted Honeyeater</figcaption>
</figure>
</div>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-2_f6b84252eadd8b472230db15e8064ea4">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"YTH"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-filter.sex.linked" class="level2">
<h2 class="anchored" data-anchor-id="run-filter.sex.linked">Run <code>filter.sex.linked</code></h2>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/ZW.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">ZW/ZZ sex chromosomes</figcaption>
</figure>
</div>
<div class="infobox task">
<p><strong>NOTE</strong></p>
<p>Set <code>ncores = 1</code> if you have less than 50,000 SNPs, since it could actually slow down the analysis.</p>
</div>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-3_a3ea2d7dd009655e505764321b3c2d55">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> YTH, <span class="at">system =</span> <span class="st">"zw"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 289 females and 347 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 46 sex-linked loci:
   2 W-linked loci
   33 sex-biased loci
   10 Z-linked loci
   1 ZW gametologs.
And 954 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-3-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="run-infer.sex" class="level2">
<h2 class="anchored" data-anchor-id="run-infer.sex">Run <code>infer.sex</code></h2>
<p>This function uses the output of function filter.sex.linked (list of 6 genlight objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an agreed sex.</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-4_6025feae0ac5070635db16fa6698f935">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">W49</td>
<td style="text-align: left;">W49</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">W90</td>
<td style="text-align: left;">W90</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">C25</td>
<td style="text-align: left;">C25</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">8</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C8</td>
<td style="text-align: left;">C8</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">7</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">W70</td>
<td style="text-align: left;">W70</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="infobox task">
<p><strong>IMPORTANT</strong></p>
<p>Any <code>agreed sex</code> with an asterisk* need to manually checked.</p>
</div>
</section>
</section>
<section id="dataset-2---xxxy---the-leadbeaters-possum" class="level1">
<h1>Dataset 2 - XX/XY - The Leadbeater’s possum</h1>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Leadbeaters_possum.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Leadbeater’s possum</figcaption>
</figure>
</div>
<section id="load-data-1" class="level2">
<h2 class="anchored" data-anchor-id="load-data-1">Load data</h2>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-5_5fc6aec291f0ed80940781169f978036">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"LBP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-filter.sex.linked-1" class="level2">
<h2 class="anchored" data-anchor-id="run-filter.sex.linked-1">Run <code>filter.sex.linked</code></h2>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/XY.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">XX/XY sex chromosomes</figcaption>
</figure>
</div>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-6_9796b72baaad9d45b0a9083c32b0087f">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> LBP, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 162 females and 211 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 13 sex-linked loci:
   0 Y-linked loci
   3 sex-biased loci
   9 X-linked loci
   1 XY gametologs.
And 987 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-6-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="run-infer.sex-1" class="level2">
<h2 class="anchored" data-anchor-id="run-infer.sex-1">Run <code>infer.sex</code></h2>
<p>This function uses the output of function filter.sex.linked (list of 6 genlight objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an agreed sex.</p>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-7_8ac66b7abfad95202f5a83a7894e2050">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough W-linked/Y-linked loci (need at least 1). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 13%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: left;">#called</th>
<th style="text-align: left;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">9</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="excercise" class="level1">
<h1>EXCERCISE</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>Choose one of the following dataset (or your own data) and report:</p>
<ol type="1">
<li><p>How many sex-linked markers are present?</p></li>
<li><p>How many individuals had a wrong sexID?</p></li>
<li><p>Do you see any changes in your PCA/structure analyses before and after filtering the sex-linked markers?</p></li>
<li><p>Do you see any differences in genetic diversity and fixation indices between autosomal and sex-linked markers?</p></li>
</ol>
</div>
</div>
<p><img src="images/beer.png" id="id" class="class" width="350" height="200"> <img src="images/Leadbeaters_possum.jpg" id="id" class="class" width="350" height="200"> <img src="images/Bull_Shark.jpg" id="id" class="class" width="350" height="200"> <img src="images/Blue_shark.jpg" id="id" class="class" width="350" height="200"></p>
<section id="exercise-data-1---your-own-data" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-1---your-own-data">Exercise data 1 - Your own data</h2>
<div class="infobox task">
<p><strong>HINT</strong></p>
<p>You can have a look at the exercise data below for inspiration.</p>
</div>
<section id="number-of-sex-linked-markers" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers">1. Number of sex-linked markers?</h3>
</section>
<section id="individuals-with-wrong-sexid" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid">2. Individuals with wrong sexID?</h3>
</section>
<section id="changes-in-pca-before-and-after-removing-the-slm" class="level3">
<h3 class="anchored" data-anchor-id="changes-in-pca-before-and-after-removing-the-slm">3. Changes in PCA before and after removing the SLM?</h3>
</section>
<section id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" class="level3">
<h3 class="anchored" data-anchor-id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</h3>
</section>
</section>
<section id="exercise-data-2---the-eastern-yellow-robin" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</h2>
<p>Data from Robledo-Ruiz et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Eastern_Yellow_Robin.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Eastern Yellow Robin</figcaption>
</figure>
</div>
<section id="load-data-2" class="level3">
<h3 class="anchored" data-anchor-id="load-data-2">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-8_fccf0f50d020e37dec20e0e83b96dce3">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"EYR"</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>EYR<span class="sc">@</span>n.loc</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>pop)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

      F   M 
  1 352 429 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-1" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-1">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-9_97f8383cc09479927ecc72e45c1b57c2">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> EYR, <span class="at">system =</span> <span class="st">"zw"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 352 females and 429 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-9-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-9-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 84 sex-linked loci:
   5 W-linked loci
   54 sex-biased loci
   18 Z-linked loci
   7 ZW gametologs.
And 916 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-9-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-1" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-1">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-10_afe8c96b480d835826bc2da9d92a4cbf">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex)  <span class="co"># remove asterisk</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(EYR<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: right;">#Hom.g</th>
<th style="text-align: right;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">13</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">15</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">12</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">11</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">14</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 11</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>Can you check if any misidentified sexes are due to uncertain genetic sexID (indicated with *)?</p>
<p><strong>HINT</strong> Try using <code>grep(pattern = "\\*", x = sexID$agreed.sex)</code></p>
</div>
</div>
</section>
<section id="changes-in-pca-before-and-after-removing-the-slm-1" class="level3">
<h3 class="anchored" data-anchor-id="changes-in-pca-before-and-after-removing-the-slm-1">3. Changes in PCA before and after removing the SLM?</h3>
<section id="pca-before" class="level4">
<h4 class="anchored" data-anchor-id="pca-before">PCA before</h4>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-11_a70c4534ec64082b8d067a29355b8c89">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>PCA.before <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.before, EYR, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.before, EYR, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
<section id="filtering" class="level4">
<h4 class="anchored" data-anchor-id="filtering">Filtering</h4>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-12_6c776c34acb9c73f26bd895eff7aaaf7">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>EYR.after <span class="ot">&lt;-</span> res<span class="sc">$</span>autosomal</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(EYR.after)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Processing genlight object with SNP data
  Reporting Read Depth by Locus
  No. of loci = 916 
  No. of individuals = 782 
    Minimum      :  2.6 
    1st quartile :  4.2 
    Median       :  5.6 
    Mean         :  6.023253 
    3r quartile  :  7.5 
    Maximum      :  12.9 
    Missing Rate Overall:  0.18 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%      12.9        1     0.1      915    99.9
2       95%      10.1       50     5.5      866    94.5
3       90%       9.3       93    10.2      823    89.8
4       85%       8.7      141    15.4      775    84.6
5       80%       8.1      185    20.2      731    79.8
6       75%       7.5      241    26.3      675    73.7
7       70%       7.1      285    31.1      631    68.9
8       65%       6.7      329    35.9      587    64.1
9       60%       6.2      379    41.4      537    58.6
10      55%       6.0      413    45.1      503    54.9
11      50%       5.6      465    50.8      451    49.2
12      45%       5.3      510    55.7      406    44.3
13      40%       5.0      551    60.2      365    39.8
14      35%       4.7      604    65.9      312    34.1
15      30%       4.4      657    71.7      259    28.3
16      25%       4.2      694    75.8      222    24.2
17      20%       4.0      738    80.6      178    19.4
18      15%       3.8      779    85.0      137    15.0
19      10%       3.5      827    90.3       89     9.7
20       5%       3.2      872    95.2       44     4.8
21       0%       2.6      916   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>EYR.after <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(EYR.after, <span class="at">lower =</span> <span class="dv">3</span>, <span class="at">upper =</span> <span class="dv">11</span>,</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.after, <span class="at">method =</span> <span class="st">"loc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data
  Reporting Call Rate by Locus
  No. of loci = 874 
  No. of individuals = 782 
    Minimum      :  0.213555 
    1st quartile :  0.7317778 
    Median       :  0.890026 
    Mean         :  0.8179257 
    3r quartile  :  0.9523655 
    Maximum      :  0.98977 
    Missing Rate Overall:  0.1821 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>EYR.after <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR, <span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">threshold =</span> <span class="fl">0.75</span>,</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data

  Reporting Call Rate by Individual
  No. of loci = 1000 
  No. of individuals = 782 
    Minimum      :  0.03 
    1st quartile :  0.814 
    Median       :  0.8435 
    Mean         :  0.8136471 
    3r quartile  :  0.86 
    Maximum      :  0.906 
    Missing Rate Overall:  0.1864 

Listing 4 populations and their average CallRates
  Monitor again after filtering
  Population CallRate   N
1     Crusoe   0.8054 238
2 Muckleford   0.8107 421
3      Timor   0.8440  52
4     Wombat   0.8368  71

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual CallRate
1    M18.29.1    0.030
2    M18.18.1    0.038
3    M18.47.2    0.055
4   027-34168    0.082
5    C18.21.2    0.083
6    C18.15.2    0.084
7    C18.16.1    0.089
8    M18.47.3    0.097
9    M18.35.2    0.134
10   M18.20.3    0.230
11   C18.28.1    0.360
12   M20.70.2    0.364
13   C18.17.2    0.390
14   C18.14.1    0.419
15   M20.70.3    0.440
16  M20.110.1    0.444
17  027-34065    0.447
18    M19.8.1    0.458
19   M20.64.3    0.483
20   M19.12.1    0.498

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>EYR.after <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.after, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">threshold =</span> <span class="fl">0.65</span>,</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(EYR.after)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
  Processing genlight object with SNP data
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Crusoe 
  No. of loci = 685 
  No. of individuals = 231 
    Minimum      :  0.0022 
    1st quantile :  0.0667 
    Median       :  0.1471 
    Mean         :  0.1823488 
    3r quantile  :  0.2822 
    Maximum      :  0.5 
    Missing Rate Overall:  0.08 

  Reporting Minor Allele Frequency (MAF) by Locus for population Muckleford 
  No. of loci = 693 
  No. of individuals = 406 
    Minimum      :  0.0012 
    1st quantile :  0.0593 
    Median       :  0.1426 
    Mean         :  0.1774444 
    3r quantile  :  0.2762 
    Maximum      :  0.4987 
    Missing Rate Overall:  0.07 

  Reporting Minor Allele Frequency (MAF) by Locus for population Timor 
  No. of loci = 603 
  No. of individuals = 52 
    Minimum      :  0.0096 
    1st quantile :  0.0673 
    Median       :  0.1667 
    Mean         :  0.1952375 
    3r quantile  :  0.305 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus for population Wombat 
  No. of loci = 641 
  No. of individuals = 71 
    Minimum      :  0.007 
    1st quantile :  0.0643 
    Median       :  0.1479 
    Mean         :  0.1821835 
    3r quantile  :  0.2803 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 715 
  No. of individuals = 760 
    Minimum      :  3e-04 
    1st quantile :  0.0606 
    Median       :  0.1383 
    Mean         :  0.1742244 
    3r quantile  :  0.272375 
    Maximum      :  0.4954 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.4954        1     0.1      709    99.9
2       95%    0.4452       36     5.1      674    94.9
3       90%    0.3799       72    10.1      638    89.9
4       85%    0.3485      107    15.1      603    84.9
5       80%    0.3092      143    20.1      567    79.9
6       75%    0.2728      178    25.1      532    74.9
7       70%    0.2371      213    30.0      497    70.0
8       65%    0.2108      249    35.1      461    64.9
9       60%    0.1902      284    40.0      426    60.0
10      55%    0.1585      320    45.1      390    54.9
11      50%    0.1379      356    50.1      354    49.9
12      45%    0.1218      391    55.1      319    44.9
13      40%    0.1018      427    60.1      283    39.9
14      35%    0.0852      462    65.1      248    34.9
15      30%    0.0722      497    70.0      213    30.0
16      25%    0.0603      533    75.1      177    24.9
17      20%    0.0488      569    80.1      141    19.9
18      15%    0.0348      605    85.2      105    14.8
19      10%    0.0210      640    90.1       70     9.9
20       5%    0.0073      675    95.1       35     4.9
21       0%    0.0003      710   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>EYR.after <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(EYR.after, <span class="at">threshold =</span> <span class="fl">0.05</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.select.colors 
  Warning: Number of required colors not specified, set to 9
  Library: RColorBrewer
  Palette: brewer.pal
  Showing and returning 2 of 9 colors for library RColorBrewer : palette Blues </code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-12-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.select.colors </code></pre>
</div>
</div>
</section>
<section id="pca-after" class="level4">
<h4 class="anchored" data-anchor-id="pca-after">PCA after</h4>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-13_1a76ee54d6dbf38bccbdc2db3fce7a5f">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>PCA.after <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR.after, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.after, EYR.after, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.after, EYR.after, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
</section>
<section id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1" class="level3">
<h3 class="anchored" data-anchor-id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm-1">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-14_20b57270b56f0a108ab68041d0c99197">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>autosomal <span class="ot">&lt;-</span> res<span class="sc">$</span>autosomal</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>z.linked <span class="ot">&lt;-</span> res<span class="sc">$</span>z.linked</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>z.linked.F <span class="ot">&lt;-</span> z.linked[z.linked<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">==</span> <span class="st">"F"</span>, ]</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>z.linked.M <span class="ot">&lt;-</span> z.linked[z.linked<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">==</span> <span class="st">"M"</span>, ]</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Genetic diversity</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>basic.autosomal <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(autosomal)</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>basic.autosomal<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.1355  0.2410  0.2520  0.0110  0.2551  0.0189  0.0438  0.0739  0.4376  0.0331 
Gst_max   Gst_H 
 0.7111  0.1040 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>basic.xF <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(z.linked.F)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>basic.xF<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.0100  0.2932  0.3052  0.0120  0.3135  0.0271  0.0393  0.0863  0.9660  0.0510 
Gst_max   Gst_H 
 0.6504  0.1327 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>basic.xM <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(z.linked.M)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>basic.xM<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.2728  0.2990  0.3144  0.0154  0.3189  0.0265  0.0488  0.0830  0.0877  0.0504 
Gst_max   Gst_H 
 0.6427  0.1292 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>divers.auto <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(autosomal, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">table =</span> <span class="cn">FALSE</span>,</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>divers.auto<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.3910761  0.3836705  0.3542339  0.3559755 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>divers.auto<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Crusoe Muckleford      Timor     Wombat
Crusoe              NA 0.03019970 0.09333132 0.06916599
Muckleford 0.007742379         NA 0.07641763 0.07214191
Timor      0.022460636 0.02616347         NA 0.09763661
Wombat     0.021589876 0.02529271 0.04001097         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>divers.zF <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(z.linked.F, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">table =</span> <span class="cn">FALSE</span>,</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>divers.zF<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.4167639  0.4723732  0.4184110  0.4361303 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>divers.zF<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor     Wombat
Crusoe             NA 0.03446323 0.11220964 0.08327249
Muckleford 0.02584944         NA 0.08811518 0.08998365
Timor      0.05283054 0.02502589         NA 0.13788524
Wombat     0.04397087 0.01616622 0.04314732         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>divers.zM <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(z.linked.M, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">table =</span> <span class="cn">FALSE</span>,</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>divers.zM<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.4145752  0.4829218  0.4104935  0.5072501 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>divers.zM<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe  Muckleford      Timor     Wombat
Crusoe             NA  0.04140262 0.11809375 0.05640494
Muckleford 0.03238801          NA 0.06872601 0.07601926
Timor      0.06860214  0.03442884         NA 0.10644965
Wombat     0.02022387 -0.01394943 0.02226470         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fixation indices</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(autosomal, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford     Timor Wombat
Crusoe             NA         NA        NA     NA
Muckleford 0.03587265         NA        NA     NA
Timor      0.05609908 0.06554101        NA     NA
Wombat     0.06413794 0.03500894 0.1013107     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(z.linked.F, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.04723621         NA         NA     NA
Timor      0.08604706 0.07861258         NA     NA
Wombat     0.09469558 0.03570803 0.01753615     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(z.linked.M, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.05611450         NA         NA     NA
Timor      0.06094081 0.08315729         NA     NA
Wombat     0.09146032 0.01952923 0.08577174     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(autosomal, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Your plot was not shown in full because your 'Plots' pane
    is too small. Increase the size of the 'Plots' pane and run the
    function again. Alternatively, use the parameter 'plot.file' to
    save the plot to a file.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0165 0.0242 0.0295
Muckleford 0.0165         NA 0.0292 0.0142
Timor      0.0242     0.0292     NA 0.0455
Wombat     0.0295     0.0142 0.0455     NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0407 0.0742 0.0795
Muckleford 0.0407         NA 0.0834 0.0474
Timor      0.0742     0.0834     NA 0.1271
Wombat     0.0795     0.0474 0.1271     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0276 0.0496 0.0528
Muckleford 0.0276         NA 0.0550 0.0304
Timor      0.0496     0.0550     NA 0.0823
Wombat     0.0528     0.0304 0.0823     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0648 0.1165 0.1247
Muckleford 0.0648         NA 0.1301 0.0736
Timor      0.1165     0.1301     NA 0.1954
Wombat     0.1247     0.0736 0.1954     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0165                      0.0242
Fstp                            0.0407                      0.0742
Dest                            0.0276                      0.0496
Gst_H                           0.0648                      0.1165
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0295                          0.0292
Fstp                        0.0795                          0.0834
Dest                        0.0528                          0.0550
Gst_H                       0.1247                          0.1301
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0142                      0.0455
Fstp                            0.0474                      0.1271
Dest                            0.0304                      0.0823
Gst_H                           0.0736                      0.1954</code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(z.linked.F, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Your plot was not shown in full because your 'Plots' pane
    is too small. Increase the size of the 'Plots' pane and run the
    function again. Alternatively, use the parameter 'plot.file' to
    save the plot to a file.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford   Timor  Wombat
Crusoe         NA     0.0211  0.0313  0.0334
Muckleford 0.0211         NA  0.0304  0.0043
Timor      0.0313     0.0304      NA -0.0154
Wombat     0.0334     0.0043 -0.0154      NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0567 0.1126 0.1234
Muckleford 0.0567         NA 0.1070 0.0644
Timor      0.1126     0.1070     NA 0.0648
Wombat     0.1234     0.0644 0.0648     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0476 0.0938 0.1058
Muckleford 0.0476         NA 0.0967 0.0582
Timor      0.0938     0.0967     NA 0.0566
Wombat     0.1058     0.0582 0.0566     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0997 0.1918 0.2141
Muckleford 0.0997         NA 0.1919 0.1169
Timor      0.1918     0.1919     NA 0.1140
Wombat     0.2141     0.1169 0.1140     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0211                      0.0313
Fstp                            0.0567                      0.1126
Dest                            0.0476                      0.0938
Gst_H                           0.0997                      0.1918
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0334                          0.0304
Fstp                        0.1234                          0.1070
Dest                        0.1058                          0.0967
Gst_H                       0.2141                          0.1919
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0043                     -0.0154
Fstp                            0.0644                      0.0648
Dest                            0.0582                      0.0566
Gst_H                           0.1169                      0.1140</code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(z.linked.F, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Your plot was not shown in full because your 'Plots' pane
    is too small. Increase the size of the 'Plots' pane and run the
    function again. Alternatively, use the parameter 'plot.file' to
    save the plot to a file.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford   Timor  Wombat
Crusoe         NA     0.0211  0.0313  0.0334
Muckleford 0.0211         NA  0.0304  0.0043
Timor      0.0313     0.0304      NA -0.0154
Wombat     0.0334     0.0043 -0.0154      NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0567 0.1126 0.1234
Muckleford 0.0567         NA 0.1070 0.0644
Timor      0.1126     0.1070     NA 0.0648
Wombat     0.1234     0.0644 0.0648     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0476 0.0938 0.1058
Muckleford 0.0476         NA 0.0967 0.0582
Timor      0.0938     0.0967     NA 0.0566
Wombat     0.1058     0.0582 0.0566     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0997 0.1918 0.2141
Muckleford 0.0997         NA 0.1919 0.1169
Timor      0.1918     0.1919     NA 0.1140
Wombat     0.2141     0.1169 0.1140     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0211                      0.0313
Fstp                            0.0567                      0.1126
Dest                            0.0476                      0.0938
Gst_H                           0.0997                      0.1918
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0334                          0.0304
Fstp                        0.1234                          0.1070
Dest                        0.1058                          0.0967
Gst_H                       0.2141                          0.1919
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0043                     -0.0154
Fstp                            0.0644                      0.0648
Dest                            0.0582                      0.0566
Gst_H                           0.1169                      0.1140</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-3---bull-shark" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-3---bull-shark">Exercise data 3 - Bull shark</h2>
<p>Data from Devloo-Delva et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Bull_Shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Bull Shark</figcaption>
</figure>
</div>
<section id="load-data-3" class="level3">
<h3 class="anchored" data-anchor-id="load-data-3">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-15_5bf9e8cf96ddc9ee3b9d232074930e4e">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Bull_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 93202</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
  85   64   33 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-2" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-2">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-16_58213f8e4ae3e57c2b05a0fb3bb42611">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 85 females and 64 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 93202.
Found 9 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   4 X-linked loci
   1 XY gametologs.
And 93193 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-16-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-2" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-2">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-17_2797d2885081a5ca66dd85c698a6290e">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex)  <span class="co"># remove asterisk</span></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CL-FIJ002-F</td>
<td style="text-align: left;">CL-FIJ002-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ003-M</td>
<td style="text-align: left;">CL-FIJ003-M</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CL-FIJ008-F</td>
<td style="text-align: left;">CL-FIJ008-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ010-F</td>
<td style="text-align: left;">CL-FIJ010-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CL-FIJ015-F</td>
<td style="text-align: left;">CL-FIJ015-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ018-F</td>
<td style="text-align: left;">CL-FIJ018-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-4---blue-shark" class="level2">
<h2 class="anchored" data-anchor-id="exercise-data-4---blue-shark">Exercise data 4 - Blue shark</h2>
<p>Data from Nikolic et al.&nbsp;2023.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Blue_shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Blue Shark</figcaption>
</figure>
</div>
<section id="load-data-4" class="level3">
<h3 class="anchored" data-anchor-id="load-data-4">Load data</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-18_8882dc0fc1ffdf051c2b36210e5bc2cc">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Blue_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 172384</code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
 104  111  149 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-3" class="level3">
<h3 class="anchored" data-anchor-id="number-of-sex-linked-markers-3">1. Number of sex-linked markers?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-19_e32626452543f0c8b761afa4a43f0f06">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">filter.sex.linked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>,</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 85 females and 64 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 93202.
Found 9 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   4 X-linked loci
   1 XY gametologs.
And 93193 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Session10_SexLinkedMarkers_files/figure-html/unnamed-chunk-19-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-3" class="level3">
<h3 class="anchored" data-anchor-id="individuals-with-wrong-sexid-3">2. Individuals with wrong sexID?</h3>
<div class="cell" data-hash="Session10_SexLinkedMarkers_cache/html/unnamed-chunk-20_0e28f2482d7687bfc477d3465cfdd290">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>,</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex)  <span class="co"># remove asterisk</span></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">60088</td>
<td style="text-align: left;">60088</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60160</td>
<td style="text-align: left;">60160</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60168</td>
<td style="text-align: left;">60168</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60176</td>
<td style="text-align: left;">60176</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60096</td>
<td style="text-align: left;">60096</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60104</td>
<td style="text-align: left;">60104</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 22</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Devloo-Delva, F., Burridge, C. P., Kyne, P. M., Brunnschweiler, J. M., Chapman, D. D., Charvet, P., … &amp; Feutry, P. (2023). From rivers to ocean basins: The role of ocean barriers and philopatry in the genetic structuring of a cosmopolitan coastal predator. Ecology and Evolution, 13(2), e9837. https://doi.org/10.1002/ece3.9837</p>
<p>Nikolic, N., Devloo-Delva, F., Bailleul, D., Noskova, E., Rougeux, C., Delord, C., … &amp; Arnaud‐Haond, S. (2023). Stepping up to genome scan allows stock differentiation in the worldwide distributed blue shark <em>Prionace glauca</em>. Molecular Ecology, 32(5), 1000-1019. https://doi.org/10.1111/mec.16822</p>
<p>Robledo-Ruiz, D. A., Austin, L., Amos, J. N., Castrejón-Figueroa, J., Harley, D. K. P., Magrath, M. J. L., Sunnucks, P., &amp; Pavlova, A. (2023). Easy-to-use R functions to separate reduced-representation genomic datasets into sex-linked and autosomal loci, and conduct sex assignment. Molecular Ecology Resources, 00, 1–21. https://doi.org/10.1111/1755-0998.13844</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>